src/components/rotate-object-button.js

```diff
@@ -0,0 +1,334 @@
+      Oc.premultiply(Cd);
+      if (exponent == 2) {
+        Oc.premultiply(Cd);
+      }
+      this.o.quaternion.copy(Oc);
+      this.o.matrixNeedsUpdate = true;
+      return;
+    }
+
+    this.o.getWorldPosition(plane.position);
+    this.camera.matrixWorld.decompose(this.v, plane.quaternion, this.v2);
+    const cameraToPlaneDistance = this.v.sub(plane.position).length();
+
+    intersections.length = 0;
+    const far = this.raycaster.far;
+    this.raycaster.far = 1000;
+    plane.raycast(this.raycaster, intersections);
+    this.raycaster.far = far;
+    const intersection = intersections[0]; // point
+    if (!intersection) return;
+
+    normal.set(0, 0, -1).applyQuaternion(plane.quaternion);
+
+    Pc.copy(intersection.point);
+    PiPc.copy(Pc).sub(Pi);
+    if (this.mode === ROTATE_MODE.FREE || this.mode === ROTATE_MODE.GARY) {
+      const boost = AFRAME.scenes[0].systems.userinput.get(paths.actions.boost);
+      this.dualAxis = !boost;
+      this.gary = boost;
+      PpPc.copy(Pc).sub(Pp);
+      XYi.copy(PiPc)
+        .projectOnPlane(normal)
+        .applyQuaternion(this.q.copy(plane.quaternion).inverse())
+        .multiplyScalar(5 / cameraToPlaneDistance);
+      if (XYi.lengthSq() < 0.00001) {
+        return;
+      }
+
+      const stepLength = Math.PI / 6;
+      this.dyAll = this.dyStore + XYi.y;
+      this.dyApplied = this.dualAxis ? Math.round(this.dyAll / stepLength) * stepLength : this.dyAll;
+      this.dyStore = this.dyAll - this.dyApplied;
+
+      this.dzAll = this.dzStore + XYi.x;
+      this.dzApplied = this.dualAxis ? Math.round(this.dzAll / stepLength) * stepLength : this.dzAll;
+      this.dzStore = this.dzAll - this.dzApplied;
+
+      this.o.matrixWorld.decompose(this.v, this.objectWorldQuaternion, this.v);
+      this.right.set(1, 0, 0).applyQuaternion(this.dualAxis ? this.objectWorldQuaternion : this.cameraWorldQuaternion);
+      this.q.setFromAxisAngle(this.right, this.dualAxis ? this.sign2 * this.sign * -this.dyApplied : -this.dyApplied);
+
+      if (this.dualAxis) {
+        this.up.set(0, 1, 0);
+      } else {
+        this.up.set(0, 1, 0).applyQuaternion(this.cameraWorldQuaternion);
+      }
+      this.q2.setFromAxisAngle(this.up, this.dzApplied);
+
+      this.o.quaternion.premultiply(this.q).premultiply(this.q2);
+    } else {
+      XYi.copy(PiPc)
+        .projectOnPlane(normal)
+        .applyQuaternion(this.q.copy(plane.quaternion).inverse())
+        .multiplyScalar(5 / cameraToPlaneDistance);
+      if (XYi.lengthSq() < 0.00001) {
+        return;
+      }
+
+      const rotationSnap = Math.PI / 8;
+      this.dzAll = this.dzStore + XYi.x;
+      this.dzApplied = Math.round(this.dzAll / rotationSnap) * rotationSnap;
+      this.dzStore = this.dzAll - this.dzApplied;
+
+      this.o.quaternion.multiply(this.q.setFromAxisAngle(this.axis, -this.sign * this.dzApplied));
+    }
+
+    Pi.copy(intersection.point);
+  }
+});
+
+AFRAME.registerComponent("exist-if-threedof", {
```

Comment:
Since all we actually care to do is set rotation mode I would call this something like `rotation-mode-selector` and check if hand poses exist each tick. If it has changed, change the `rotation-button` mode.